engine/spec/components/input_spec.rb

```diff
@@ -0,0 +1,137 @@
+  let(:error_message) { nil }
+  let(:optional) { false }
+  let(:hint) { nil }
+  let(:additional_attributes) { nil }
+  let(:width) { nil }
+
+  context "by default" do
+    it "renders the label" do
+      expect(subject.text.strip).to include("Example input")
+    end
+
+    it "links the label to the input" do
+      expect(subject.css("label").attribute("for").value).to eq("example-input-input")
+    end
+
+    it "renders an input" do
+      expect(subject.css("input")).to be_present
+    end
+
+    it "adds the correct id to the input" do
+      expect(subject.css("input").attribute("id").value).to eq("example-input-input")
+    end
+
+    it "renders a text input" do
+      expect(subject.css("input").attribute("type").value).to eq("text")
+    end
+
+    it "renders an empty input" do
+      expect(subject.css("input").attribute("value")).to eq nil
+    end
+
+    it "renders a required input" do
+      expect(subject.css("input").attribute("required")).to be_present
+    end
+  end
+
+  context "when an error is present" do
+    let(:error_message) { "Enter your name" }
+
+    it "renders the error message" do
+      expect(subject.text.strip).to include "Enter your name"
+    end
+
+    it "adds aria-invalid to the input" do
+      expect(subject.css("input").attribute("aria-invalid")).to be_present
+    end
+
+    it "adds aria-describedby to the input" do
+      expect(subject.css("input").attribute("aria-describedby").value).to eq("example-input-error")
+    end
+  end
+
+  context "when the input is optional" do
+    let(:optional) { true }
+
+    it "adds optional to the label" do
+      expect(subject.text).to include("Optional")
+    end
+
+    it "does not add required to the input" do
+      expect(subject.css("input").attribute("required")).to_not be_present
+    end
+  end
+
+  context "when there is hint text" do
+    let(:hint) { "This is the hint text" }
+
+    it "renders the hint text" do
+      expect(subject.text.strip).to include("This is the hint text")
+    end
+  end
+
+  context "when a type is specified" do
+    let(:type) { "email" }
+
+    it "renders the correct type of input" do
+      expect(subject.css("input").attribute("type").value).to eq("email")
+    end
+  end
+
+  context "when additional attributes are specified" do
+    let(:additional_attributes) do
+      {
+        autocomplete: "name",
+        "data-foo": "bar"
+      }
+    end
+
+    it "adds the attributes to the input" do
+      expect(subject.css("input").attribute("autocomplete").value).to eq("name")
+      expect(subject.css("input").attribute("data-foo").value).to eq("bar")
+    end
+  end
+
+  context "when a valid width is specified" do
+    let(:width) { "4ch" }
+
+    it "renders the input at the correct width" do
+      expect(subject.css(".cads-input--4ch")).to be_present
+    end
+  end
+
+  context "when an invalid width is specified" do
+    let(:width) { "banana" }
+
+    context "production rails env" do
```

Comment:
from a testing perspective there probably should be a non-production unit test also.

You can also get hints for gaps like this by using something like `simplecov` I'm not sure if we use it in public-website. It's a code-coverage gem that's reasonably simple to setup.

If you change the below line to         `allow(Rails.env).to receive(:production?).and_return(production?)` then you can then specify the flag switch in the same way you do other items.

Reply:
This is really testing a single helper https://github.com/citizensadvice/design-system/blob/master/engine/app/lib/citizens_advice_components/fetch_or_fallback_helper.rb so it would be nice to cover this in a what that doesn't feel like we need the same boilerplate every time. Any suggestions for using shared examples or similar approaches here?

Reply:
Could we test the fallback helpers directly and then not test them in the components?  Feels like the coverage would be there?  

Would be keen to know what you think though @luke-hill - basically what is the best way to test that invalid params are handled correctly? (as defined by these helpers here: https://github.com/citizensadvice/design-system/blob/master/engine/app/lib/citizens_advice_components/fetch_or_fallback_helper.rb and implemented by the components here: https://github.com/citizensadvice/design-system/blob/8665e1715ae977b8b81e64cabfa2aaa4ca276e05/engine/app/components/citizens_advice_components/callout.rb#L9)

Reply:
Yep, it's quite common to write helper tests. If this is just consuming that.

The theory behind simplecov (If you are interested), is just each LOC is tested once.

If you want to test the helper a good way (My recommendation is).

```
let(:dummy) do
  Class.new do
    include MyHelper
  end
end

subject(:named_subject) { dummy.new }


context "when good" do
  it "passes" do
    expect(dummy).to be_awesome
  end
end

context "when crap" do
  it "fails" do
    expect(dummy).to raise_error(ArgumentError).with_message("Be better!")
  end
end
```

Reply:
EDIT: When testing them, the folder structure should mimic the lib. So the tests would probably live in `spec/helpers/myhelper_spec.rb`

Reply:
Perfect. In fact it looks like Primer—where we pinched `fetch_or_fallback` from—does a similar thing albeit in minitest.

Reply:
Raised this https://github.com/citizensadvice/design-system/pull/1273. Thanks for the nudge in the right direction!